exclude config option

shahzebsiddiqui opened the pull request:
This PR will address #267 
- The option is ``-x`` and long option ``--exclude``
- The exclude can be passed multiple times and it can be a file or directory or both and works with shell expansion
- If -x is not specified then nothing to exclude
- Exclusion happens after configs are discovered
- If no configs are to be processed after exclusion then buildtest will stop since nothing is to be done
- Exclusion happens after ``discover_configs``
- There is a regtest to cover exclude
- Also updated a few example configs so that all test run to completion whether they PASS/FAIL 

Here are few examples to illustrate how it works

Undo detection example
```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c examples -x examples/
There are no config files to process.
```

Example removing two configs  file from exclusion
```

(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c examples/script/ -x examples/script/bzip2.yml -x examples/script/slurm.yml
Config Name                    SubTest                        Status                         Config Path              
________________________________________________________________________________________________________________________
python-circle                  circle_area                    PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/python-circle.yml
zlib                           zlib_compression               PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/zlib.yml
zlib                           zlib_decompress                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/zlib.yml
==============================================================
                         Test summary
Executed 3 tests
Passed Tests: 3/3 Percentage: 100.0%
Failed Tests: 0/3 Percentage: 0.0%
```

Exclude by directory in this case everything is skipped under ``examples/script`` 

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c examples/ -x examples/script/
Config Name                    SubTest                        Status                         Config Path              
________________________________________________________________________________________________________________________
stream                         stream_uniprocess_c            PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
stream                         stream_openmp_c                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
==============================================================
                         Test summary
Executed 2 tests
Passed Tests: 2/2 Percentage: 100.0%
Failed Tests: 0/2 Percentage: 0.0%

```


Here is an example where one directory and file is passed. Notice in this case we exclude the whole benchmark sub-directory  in this case only contains stream test and we exclude bzip2 as well.

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c examples/ -x examples/benchmarks/ -x examples/script/bzip2.yml
Config Name                    SubTest                        Status                         Config Path              
________________________________________________________________________________________________________________________
python-circle                  circle_area                    PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/python-circle.yml
zlib                           zlib_compression               PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/zlib.yml
zlib                           zlib_decompress                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/zlib.yml
slurm                          slurm_down_nodes_reason        FAILED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
slurm                          slurm_not_responding_nodes     FAILED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
==============================================================
                         Test summary
Executed 5 tests
Passed Tests: 3/5 Percentage: 60.0%
Failed Tests: 2/5 Percentage: 40.0%

```

shahzebsiddiqui committed:
add option to exclude config this is a new option "buildtest build -x"
There is a long option --exclude
The exclude option can be passed multiple times to exclude more than one
file or directory.
After discover stage, the exclusions is applied if no files to be
processed after exclusion then buildtest will terminate since there is
no work to be done.
In discover stage we get abspath of all files that way exclusion works
correctly.
Add first attempt to test exclude_configs method. If no files are passed
for exclusion (i.e no -x provided) the normalized configs will be same
as the ones detected in discover stage.

shahzebsiddiqui committed:
add example directory example for excluding configs.

vsoch commented on buildtest/menu/build.py:
```diff
@@ -74,10 +74,65 @@ def discover_configs(config_file):
         sys.exit(msg)
 
+    # return all configs with absolute path
+    tmp = [os.path.abspath(config) for config in config_files]
```
This is where you'd want to put a more elegant util function to handle returning absolute, symlink resolved paths.

vsoch replied:
Also we should try to avoid a variable name like `tmp` if possible, there is usually a better way to do it.

shahzebsiddiqui replied:
Okay i was able to resolve this and the bug i posted about symlinks by using os.realpath https://docs.python.org/2/library/os.path.html#os.path.realpath which resolves symlinks.

Now this test is working as expected. 
 ```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c examples/benchmarks/  -x $PWD/examples/benchmarks
There are no config files to process.
```

My symlink gpfs is now resolving any any subdirectory in symlink works too
```
>>> os.path.realpath('/u/users/ssi29/gpfs')
'/mxg-hpc/users/ssi29'
>>> os.path.realpath('/u/users/ssi29/gpfs/buildtest-framework')
'/mxg-hpc/users/ssi29/buildtest-framework'
```

What won't work is  shell expansion is required then realpath won't expand automatically

```
>>> os.path.realpath('$HOME/gpfs/buildtest-framework')
'/mxg-hpc/users/ssi29/buildtest-framework/$HOME/gpfs/buildtest-framework'
```

But doing shell expansion first then doing realpath would work

```
>>> os.path.realpath(os.path.expandvars('$HOME/gpfs/buildtest-framework'))
'/mxg-hpc/users/ssi29/buildtest-framework'
```

Then it comes to the other odd case of ``~`` if its used when referring to file, so not sure if ``resolve_path`` method should take care of all of these cases in the following order
- Run os.path.expandvars address ``$HOME``
- Run os.path.expandusers address ``~``
- Run os.path.realpath to resolve symlinks
- Finally run ``os.path.exists`` to be absolutely sure its a path.

vsoch replied:
You don't need the tmp here, you can have an iterator that returns to a variable with the same name (not like a dictionary that you can't update in place).
```python
config_files = [os.path.abspath(config) for config in config_files]
```

vsoch commented on buildtest/menu/build.py:
```diff
@@ -74,10 +74,65 @@ def discover_configs(config_file):
     return config_files
 
 
+def exclude_configs(config_files, exclude_list):
```
This is way overly complicated a function for such a simple action! We basically need to get the resolved paths for all discovered configs, create a single regular expression that combines the exclude patterns, and then (in one loop) go over them and keep ones not excluded. Here is an example of a function I wrote that does the same thing in just a few lines: https://github.com/urlstechie/urlchecker-python/blob/04c84629eabaa70a131a670308eb54c741d62cf2/urlchecker/core/fileproc.py#L37

And then is used in a single iterator https://github.com/urlstechie/urlchecker-python/blob/04c84629eabaa70a131a670308eb54c741d62cf2/urlchecker/core/fileproc.py#L82

shahzebsiddiqui replied:
When i was implementing ``--exclude`` i didn't intend for regular expression. Regex would be nice but i foresee that to be a significant change or maybe even a new option. Right now i mimic ``-c`` with ``-x`` which supports file or directory. Both of them don't support regex. 

I think advanced users that know regex would prefer a pattern search for both ``-c`` and ``-x``, but most novice users would prefer just passing a file or directory.

Would it make sense to stand up a new option for regular expression, or we can support regex into both ``-c`` and ``-x`` to be consistent. One thing that i realized was ``-c`` can't be used twice which i can change by adding action ``append`` so something like

```
buildtest build -c dir1 -c dir2
```

vsoch replied:
The --exclude doesn't take a regexp, it turns into a list, and that would be parsed into a regular expression. E.g.,:

```python
exclude = ["dir1", "dir2"]
```
turns into

```python
"(dir1|dir2)"
```
The command line interface you have does not change.

shahzebsiddiqui replied:
I see the regex examples i am not entirely if regex will do trick or my lack of understanding. For a directory i need to retrieve all yml files and add to exclude list before checking if they are already discovered so i can remove them.

vsoch replied:
Maybe this will help:

 1. You start with a list of files, config_files
 2. You run them through an iterate to filter that determines if they should be included
 3. The filter takes the entire list of excludes and generates _one_ regular expression to check the _one_ config file path that is being passed over
 4. The function returns True or False to determine to include or not. E.g.,

```
exclude_list = ["dir1", "dir2"]
config_files = [config for config in config_files if include_file(config, exclude_list)]
```
And the include file function is similar to what I showed you.

shahzebsiddiqui replied:
Okay i fixed this now and i was able to fix resolve path to support both shell and user expansion and applied resolve path to config files too. So this two tests are now discovered properly

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c $PWD/examples
Config Name                    SubTest                        Status                         Config Path              
________________________________________________________________________________________________________________________
python-circle                  circle_area                    PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/python-circle.yml
zlib                           zlib_compression               PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/zlib.yml
zlib                           zlib_decompress                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/zlib.yml
bzip2                          bzip_compression               PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/bzip2.yml
bzip2                          bzip_help                      PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/bzip2.yml
bzip2                          bzip2_help                     PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/bzip2.yml
slurm                          slurm_down_nodes_reason        FAILED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
slurm                          slurm_not_responding_nodes     FAILED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
stream                         stream_uniprocess_c            PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
stream                         stream_openmp_c                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
==============================================================
                         Test summary
Executed 10 tests
Passed Tests: 8/10 Percentage: 80.0%
Failed Tests: 2/10 Percentage: 20.0%
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c ~/gpfs/buildtest-framework/examples
Config Name                    SubTest                        Status                         Config Path              
________________________________________________________________________________________________________________________
python-circle                  circle_area                    PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/python-circle.yml
zlib                           zlib_compression               PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/zlib.yml
zlib                           zlib_decompress                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/zlib.yml
bzip2                          bzip_compression               PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/bzip2.yml
bzip2                          bzip_help                      PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/bzip2.yml
bzip2                          bzip2_help                     PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/bzip2.yml
slurm                          slurm_down_nodes_reason        FAILED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
slurm                          slurm_not_responding_nodes     FAILED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
stream                         stream_uniprocess_c            PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
stream                         stream_openmp_c                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
==============================================================
                         Test summary
Executed 10 tests
Passed Tests: 8/10 Percentage: 80.0%
Failed Tests: 2/10 Percentage: 20.0%

```

shahzebsiddiqui replied:
Thanks @vsoch  the solution does work i had to make a minor adjustment to apply ``resolve_path`` before running regex since paths needed to be resolved for regex to work, i had applied some logging to show some details.

Here is an example with user expansion
 
```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c ~/gpfs/buildtest-framework/examples -x ~/gpfs/buildtest-framework/examples/
There are no config files to process.
```

The logfile will have the following
```
2020-04-10 09:38:39,875 [build.py:155 - func_build_subcmd() ] - [DEBUG] The exclude config pattern are the following: -e ['/u/users/ssi29/gpfs/buildtest-framework/examples/']
2020-04-10 09:38:39,876 [build.py:105 - include_file() ] - [DEBUG] white list pattern before resolving paths: ['/u/users/ssi29/gpfs/buildtest-framework/examples/']
2020-04-10 09:38:39,876 [build.py:109 - include_file() ] - [DEBUG] white list pattern after resolving paths: ['/mxg-hpc/users/ssi29/buildtest-framework/examples']
2020-04-10 09:38:39,876 [build.py:112 - include_file() ] - [DEBUG] Applying Regular Expression Search: (/mxg-hpc/users/ssi29/buildtest-framework/examples) to file: /mxg-hpc/users/ssi29/buildtest-framework/examples/script/python-circle.yml
```

Here is an example with shell expansion where only ``examples/benchmark`` is supposed to be processed

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c examples/ -x $HOME/gpfs/buildtest-framework/examples/script
Config Name                    SubTest                        Status                         Config Path              
________________________________________________________________________________________________________________________
stream                         stream_uniprocess_c            PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
stream                         stream_openmp_c                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
==============================================================
                         Test summary
Executed 2 tests
Passed Tests: 2/2 Percentage: 100.0%
Failed Tests: 0/2 Percentage: 0.0%
```

This example just shows we pass multiple exclusion via ``-x``  which works by filtering output as expected

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -c examples/ -x examples/script/bzip2.yml -x examples/script/zlib.yml
Config Name                    SubTest                        Status                         Config Path              
________________________________________________________________________________________________________________________
python-circle                  circle_area                    PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/python-circle.yml
slurm                          slurm_down_nodes_reason        FAILED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
slurm                          slurm_not_responding_nodes     FAILED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/script/slurm.yml
stream                         stream_uniprocess_c            PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
stream                         stream_openmp_c                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
==============================================================
                         Test summary
Executed 5 tests
Passed Tests: 3/5 Percentage: 60.0%
Failed Tests: 2/5 Percentage: 40.0%
```

And the logfile will have the following regex

```
2020-04-10 09:45:06,706 [build.py:112 - include_file() ] - [DEBUG] Applying Regular Expression Search: (/mxg-hpc/users/ssi29/buildtest-framework/examples/script/bzip2.yml|/mxg-hpc/users/ssi29/buildtest-framework/examples/script/zlib.yml) to file: /mxg-hpc/users/ssi29/buildtest-framework/examples/script/zlib.yml
```

vsoch commented on examples/script/bzip2.yml:
```diff
@@ -11,12 +11,8 @@ bzip_compression:
     tar cvjf "$TARFILE" "$FILE"
     ls -l "$TARFILE"
-
-bzip_version:
```
Why did you remove this test?

shahzebsiddiqui replied:
it was getting a non-zero exit and was failing

vsoch commented on examples/script/zlib.yml:
```diff
@@ -2,18 +2,28 @@ version: 0.0.1
-    ls -lh 1mb.tar.gz
-    gzip -l 1mb.tar.gz
+    dd if=/dev/zero of=$FILE bs=1M count=1
+    tar cvzf $TARFILE $FILE
```
Is there a reason you don't quote bash variables or put them in `{}`? I thought that was considered best practice.

shahzebsiddiqui replied:
that's a typo on my end i will use ``{}`` i can fix this

shahzebsiddiqui committed:
add method resolve_path this takes care of retrieving full canonical
path and resolve symlinks. We also take care of shell expansion.
Update test examples to use {} when referencing shell variables

vsoch commented on buildtest/utils/file.py:
```diff
@@ -119,3 +119,22 @@ def create_dir(dirname):
+    real_path = os.path.realpath(shell_expansion)
+    if os.path.exists(real_path):
+        return real_path
+    else:
```
A function will by default return None if nothing is returned, so you don't need this else -> return None.

shahzebsiddiqui replied:
that was a dumb mistake good catch

shahzebsiddiqui replied:
This is fixed now

shahzebsiddiqui committed:
fix issue related by CodeFactor

shahzebsiddiqui committed:
refactor how exclusion is done by running regular expression. Add
logging for discovery and exclusion stage.
Resolve paths during discovery and exclusion so they work with
shell and user expansion.

shahzebsiddiqui committed:
change output format by adding list of discovered files.
Fix floating point result to 3 digits on pass and fail rate.

shahzebsiddiqui committed:
fix issue reported by black.

vsoch commented on buildtest/menu/build.py:
```diff
@@ -74,19 +74,57 @@ def discover_configs(config_file):
 
+def include_file(file_path, white_list_patterns):
+    """Check if file is included based on OR regular expression. If no white_list_patterns
+       provided method will return ``True``. Otherwise it will return the list of files that don't
```
so much cleaner! <3

vsoch commented on buildtest/menu/build.py:
```diff
@@ -74,19 +74,57 @@ def discover_configs(config_file):
+       :rtype: bool or str
+    """
+
+    logger = logging.getLogger(__name__)
```
Is there a reason that the logger is instantiated here, instead of for the module?

shahzebsiddiqui replied:
ahh you mean just instantiate once per file as global variable and not call inside each method. I think that makes sense. If that's what you mean i think it makes sense and i agree with your logic.

vsoch replied:
Exactly - that's why we use __name__ for the logger, it refers to the active module.

shahzebsiddiqui replied:
I fixed this now

vsoch commented on buildtest/menu/build.py:
```diff
@@ -74,19 +74,57 @@ def discover_configs(config_file):
+    logger.debug(f"white list pattern after resolving paths: {white_list_patterns}")
+
+    regexp = "(%s)" % "|".join(white_list_patterns)
+    logger.debug(f"Applying Regular Expression Search: {regexp} to file: {file_path}")
```
This is probably one too many lines for the logger - is there a lower level than debug (e.g., verbose) for at least the last line?

shahzebsiddiqui replied:
DEBUG is the lowest level i believe according to https://docs.python.org/3/howto/logging.html We could either remove the line if it's too much detail or we can set default debug setting to INFO. The default is ``WARNING``.

I figured it could be useful for users to troubleshoot how exclusion is happening if they have information on the regex command.

vsoch replied:
ah okay, then okay to leave it. I think this is why I implemented a custom logger for most of my libraries that has also logger.verbose().

vsoch commented on buildtest/menu/build.py:
```diff
@@ -111,6 +149,35 @@ def func_build_subcmd(args):
+    # if no files discovered let's stop now
+    if not config_files:
+        msg = "There are no config files to process."
+        logger.error(msg)
```
What is the reason for wanting the message to go to the logger.error _and_ system.exit?

shahzebsiddiqui replied:
Good point, i don't have any particular reasoning. I guess it doesn't make sense to have same error message if printed to screen. I guess it would be only useful if an exception message is captured before exiting for detailed error message. That's what i am accustomed to, just ``tail``  the last few lines to get the error message.

vsoch replied:
I was thinking it would be cleaner to do something like:

```python
logger.error("There was an error")  # <--- error message goes to log
sys.exit(1)                                       # <--- exit with error code
```

shahzebsiddiqui replied:
I will remove this line and fixed locally.

vsoch commented on buildtest/menu/build.py:
```diff
@@ -111,6 +149,35 @@ def func_build_subcmd(args):
+
+    print("\n")
+    print("{:^45}".format("Discovered Files"))
+    print("\n")
```
Can we put the newlines with other statements so this single block doesn't need to take up 7 lines?

shahzebsiddiqui replied:
Makes sense i can make this adjustment

vsoch commented on buildtest/menu/build.py:
```diff
@@ -111,6 +149,35 @@ def func_build_subcmd(args):
+    if args.exclude:
+        logger.debug(f"The exclude config pattern are the following: -e {args.exclude}")
+
+    config_files = [
```
Is this needed to run if args.exclude isn't set? Shouldn't it be under the if args.exclude?

shahzebsiddiqui replied:
that would be most appropriate i will make this fix.

vsoch commented on buildtest/menu/build.py:
```diff
@@ -111,6 +149,35 @@ def func_build_subcmd(args):
+        config for config in config_files if include_file(config, args.exclude)
+    ]
+
+    logger.debug(
```
What do you mean normalized?

shahzebsiddiqui replied:
Bad naming, what about *filtered*, need some way to denote that configs discovered after exclusion are the ones that buildtest will process.

vsoch commented on buildtest/menu/build.py:
```diff
@@ -147,16 +214,22 @@ def func_build_subcmd(args):
-        print("                         Test summary                         ")
+        print("\n")
+        print("\n")
+        print("{:=<60}".format(""))
```
Let's also combine some of these statements, you can add the \n to have more than one per print().

shahzebsiddiqui replied:
makes sense. Fixed now

vsoch commented on buildtest/utils/file.py:
```diff
@@ -119,3 +119,24 @@ def create_dir(dirname):
+       :return: return realpath to file if found otherwise return None
+       :rtype: str or None
+    """
+    # apply shell expansion  when file includes something like $HOME/example
```
Great job on documentation here!

vsoch commented on buildtest/utils/file.py:
```diff
@@ -119,3 +119,24 @@ def create_dir(dirname):
+
+
+def resolve_path(path):
+    """This method will resolve a file path with accounts for shell expansion and resolve paths in
```
This method will resolve a file path _to account for_ shell expansion...

shahzebsiddiqui replied:
fixed locally

vsoch commented on examples/script/bzip2.yml:
```diff
@@ -8,15 +8,11 @@ bzip_compression:
+    dd if=/dev/zero of=${FILE} bs=1M count=1
+    tar cvjf ${TARFILE} ${FILE}
+    ls -l ${TARFILE}
+    rm $TARFILE
```
Nit: be consistent in using the {} for the lsat line as well.

shahzebsiddiqui replied:
fixed locally good catch.

shahzebsiddiqui committed:
move logger as global variable in buildtest/menu/build.py
Condense some print statements into single line to make code condense.
Move exclusion check under if condition for args.exclude
Fix bzip2.yml test with missing {}

vsoch commented on buildtest/menu/build.py:
```diff
@@ -152,31 +149,22 @@ def func_build_subcmd(args):
-    print("\n")
+    print("\n {:^45} \n".format("Discovered Files"))
+    [print(config) for config in config_files]
+    print("\n\n")
```
Much more succinct.
